### PR TITLE
Fix to issue 724 -- sendmail after password change or password reset

### PR DESCRIPTION
**Pull request by unithiyashree**
### Proposed changes in this pull request

- [This change fixes the issue #724 "Send an email when a user changes the password"]

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Risks

[List any risks that could arise from merging your pull request.]

-

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: cadasta/accounts/tests/test_forms.py [hunk 1]
```diff
@@ -60,20 +60,6 @@ def test_password_contains_username(self):
                 form.errors.get('password1'))
         assert User.objects.count() == 0
 
-    def test_password_contains_blank_username(self):
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: Why do you want to delete this test?

File: cadasta/accounts/tests/test_forms.py [hunk 2]
```diff
@@ -89,20 +75,6 @@ def test_password_contains_email(self):
                 form.errors.get('password1'))
         assert User.objects.count() == 0
 
-    def test_password_contains_blank_email(self):
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: Why do you want to delete this test?

File: cadasta/accounts/tests/test_forms.py [hunk 3]
```diff
@@ -282,7 +254,20 @@ def test_valid_data(self):
 
         assert User.objects.count() == 1
 
-        assert user.check_password('iloveyoko79!') is True
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: Why do you want to remove this line from the test?

File: cadasta/accounts/tests/test_forms.py [hunk 4]
```diff
@@ -282,7 +254,20 @@ def test_valid_data(self):
...
+            'oldpassword': 'beatles4L1yfe!',
+            'password1': 'iloveyoko719!',
+            'password2': 'iloveyoko719!',
+        }
+
+        form = forms.ChangePasswordForm(user, data)
+        assert form.is_valid() is True
+        form.save()
+
+        self.assertEqual(len(mail.outbox), 1)
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: Please use `assert len(mail.outbox) == 1` instead of `self.assertEqual` to be consistent with the rest of our tests.

File: cadasta/accounts/tests/test_forms.py [hunk 5]
```diff
@@ -392,6 +377,20 @@ def test_valid_data(self):
...
+        data = {
+            'password1': 'iloveyoko719!',
+            'password2': 'iloveyoko719!',
+        }
+
+        form = forms.ResetPasswordKeyForm(data, user=user)
+        assert form.is_valid() is True
+        form.save()
+
+        self.assertEqual(len(mail.outbox), 1)
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: Please use `assert len(mail.outbox) == 1` instead of `self.assertEqual` to be consistent with the rest of our tests.

File: cadasta/accounts/forms.py [hunk 6]
```diff
@@ -122,9 +123,25 @@ def clean_password1(self):
 
 class ChangePasswordForm(ChangePasswordMixin,
                          allauth_forms.ChangePasswordForm):
-    pass
+    def save(self, *args, **kwargs):
+        super(ChangePasswordForm, self).save(*args, **kwargs)
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: We're using Python 3, you can simply write `super().save(*args, **kwargs)`

File: cadasta/accounts/forms.py [hunk 7]
```diff
@@ -122,9 +123,25 @@ def clean_password1(self):
...
+            [self.user.email],
+            fail_silently=False,
+        )
 
 
 class ResetPasswordKeyForm(ChangePasswordMixin,
                            allauth_forms.ResetPasswordKeyForm):
-    pass
+    def save(self, *args, **kwargs):
+        super(ResetPasswordKeyForm, self).save(*args, **kwargs)
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: We're using Python 3, you can simply write `super().save(*args, **kwargs)`

File: cadasta/accounts/forms.py [hunk 8]
```diff
@@ -122,9 +123,25 @@ def clean_password1(self):
 
 class ChangePasswordForm(ChangePasswordMixin,
                          allauth_forms.ChangePasswordForm):
-    pass
+    def save(self, *args, **kwargs):
+        super(ChangePasswordForm, self).save(*args, **kwargs)
+        send_mail(
+            "Password Changed",
+            "you have successfully changed password",
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: I would suggest to use a more informative text here, for example: "You are receiving this email because someone tried to change the password for your account at Cadasta Platform. Please contact us immediately under security (at) cadasta (dot) org if you have not changed your password."

I would also suggest to use an email template instead of hard-coding the message and marking the text of the email for translation so we send out the email in the user's preferred language.

File: cadasta/accounts/forms.py [hunk 9]
```diff
@@ -122,9 +123,25 @@ def clean_password1(self):
...
 
 
 class ResetPasswordKeyForm(ChangePasswordMixin,
                            allauth_forms.ResetPasswordKeyForm):
-    pass
+    def save(self, *args, **kwargs):
+        super(ResetPasswordKeyForm, self).save(*args, **kwargs)
+        send_mail(
+            "Password Changed",
+            "you have successfully changed password",
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: I would suggest to use a more informative text here, for example: "You are receiving this email because someone tried to change the password for your account at Cadasta Platform. Please contact us immediately under security (at) cadasta (dot) org if you have not changed your password."

I would also suggest to use an email template instead of hard-coding the message and marking the text of the email for translation so we send out the email in the user's preferred language.

File: cadasta/accounts/forms.py [hunk 10]
```diff
@@ -122,9 +123,25 @@ def clean_password1(self):
 
 class ChangePasswordForm(ChangePasswordMixin,
                          allauth_forms.ChangePasswordForm):
-    pass
+    def save(self, *args, **kwargs):
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: You're using similar functionality in both `ChangePasswordForm` and `ResetPasswordKeyForm`. How about moving the `save` method into `ChangePasswordMixin` because both Forms extend the mixin?

File: cadasta/accounts/tests/test_forms.py (hunk 1 above)
Comment (unithiyashree) [inline cadasta/accounts/tests/test_forms.py]: sorry @oliverroick  I have unfortunately deleted this test function. I will make the changes and submit the correct one.

File: cadasta/accounts/tests/test_forms.py (hunk 2 above)
Comment (unithiyashree) [inline cadasta/accounts/tests/test_forms.py]: sorry @oliverroick I have unfortunately deleted this test function. I will make the changes and submit the correct one.

File: cadasta/accounts/tests/test_forms.py (hunk 3 above)
Comment (unithiyashree) [inline cadasta/accounts/tests/test_forms.py]: sorry @oliverroick I have unfortunately deleted this line. I will make the changes and submit the correct one.

File: cadasta/accounts/tests/test_forms.py (hunk 4 above)
Comment (unithiyashree) [inline cadasta/accounts/tests/test_forms.py]: thank you @oliverroick  for guiding me. I will use assert statement as mentioned.

File: cadasta/accounts/tests/test_forms.py (hunk 5 above)
Comment (unithiyashree) [inline cadasta/accounts/tests/test_forms.py]: thank you @oliverroick for guiding me. I will use assert statement as mentioned.

File: cadasta/templates/allauth/account/email/password_changed_message.txt [hunk 11]
```diff
@@ -0,0 +1,6 @@
+{% load i18n %}{% blocktrans %}You are receiving this email because someone changed the password for your account at Cadasta Platform. {% endblocktrans %}
+
+{% blocktrans %}Please contact us immediately under security (at) cadasta (dot) org if you have not changed your password.{% endblocktrans %}
+
+{% blocktrans with site_name=current_site.name site_domain=current_site.domain %}Thank you for using Cadasta Platform!
```
Comment (oliverroick) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: `current_site` is not in the template context, so you can remove `with site_name=current_site.name site_domain=current_site.domain`

File: cadasta/accounts/forms.py [hunk 12]
```diff
@@ -119,6 +121,25 @@ def clean_password1(self):
 
         return password
 
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+        msg_subject = render_to_string(
+                'account/email/'
```
Comment (oliverroick) [inline cadasta/accounts/forms.py]: The template path easily fits into one line. Please change that because it's more readable that way. You could also remove four white-spaces at the beginning of the line.

File: cadasta/templates/allauth/account/email/password_changed_message.txt [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{% load i18n %}{% blocktrans %}You are receiving this email because someone changed the password for your account at Cadasta Platform. {% endblocktrans %}
```
Comment (oliverroick) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: Can you move the template into `/accounts/email/`? It's a custom email, we don't overwrite allauth's standard emails here.

File: cadasta/templates/allauth/account/email/password_changed_subject.txt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
```
Comment (oliverroick) [inline cadasta/templates/allauth/account/email/password_changed_subject.txt]: Can you move the template into `/accounts/email/`? It's a custom email, we don't overwrite allauth's standard emails here.

File: cadasta/templates/allauth/account/email/password_changed_subject.txt [hunk 15]
```diff
@@ -0,0 +1,4 @@
...
+{% autoescape off %}
+{% blocktrans %}Password Changed Email{% endblocktrans %}
```
Comment (oliverroick) [inline cadasta/templates/allauth/account/email/password_changed_subject.txt]: Can you reword this to "Your password at Cadasta was changed"

File: cadasta/templates/allauth/account/email/password_changed_message.txt (hunk 13 above)
Comment (oliverroick) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: I would put the whole message into one `{% blocktrans %}` block. Otherwise translators will find three individual  snippets of the whole email, which will make it more difficult to translate consistently.

File: cadasta/templates/allauth/account/email/password_changed_subject.txt (hunk 15 above)
Comment (unithiyashree) [inline cadasta/templates/allauth/account/email/password_changed_subject.txt]: @oliverroick  Thank you for suggesting the changes.

File: cadasta/templates/allauth/account/email/password_changed_subject.txt (hunk 14 above)
Comment (unithiyashree) [inline cadasta/templates/allauth/account/email/password_changed_subject.txt]: @oliverroick Thank you for suggesting the changes.

File: cadasta/templates/allauth/account/email/password_changed_message.txt (hunk 11 above)
Comment (unithiyashree) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: @oliverroick Thank you for suggesting the changes.

File: cadasta/templates/allauth/account/email/password_changed_message.txt (hunk 13 above)
Comment (unithiyashree) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: @oliverroick Thank you for suggesting the changes.

Comment (unithiyashree) [inline cadasta/templates/allauth/account/email/password_changed_message.txt]: @oliverroick Thank you for suggesting the changes.

**Comment by unithiyashree** [comment]
@oliverroick Thank you. I am sorry forgot to delete the files from alluth folder. Please review the PR.

**Comment by unithiyashree** [comment]
Hi @oliverroick .. Can you please review the PR? All the changes have been made according to your suggestions.

**Comment by unithiyashree** [comment]
@oliverroick  @linzjax  thank you for reviewing the PR.

**Comment by oliverroick** [comment]
Hi @unithiyashree, we tried to merge your PR today, but we have not been able to rebase your branch into the latest changes in `master`.

Your commit history seems a bit off, there are, for example, several identical commits. Could you please have a look at this and try to rebase your branch?

You'll see that GitHub says "This branch is out-of-date with the base branch". That means that there have been changes in the main Cadasta/cadasta-platform repository that you don't have in your PR branch. That means that there's no way to know if your work is compatible with what's already been done. What you need to do in this case is to make sure that you've added our repo as an upstream remote for your fork (GitHub usually tells you to do this when you create a fork), fetch the latest master from there, and then update your PR branch to incorporate the latest changes. (This is the point where working directly on master in your repo turns out to be a bad idea...) If you're working on a topic branch, you do this:

```
git checkout master  # This is your local master branch
git fetch upstream/master  # Get changes from Cadasta/cadasta-platform
git merge upstream/master  # This does a "fast-forward" merge
git checkout <topic-branch>  # Switch back to your topic branch
git rebase master  # Replay the changes you've made on top of the latest upstream master
git push --force  # Update your PR on GitHub
```

If there are conflicts between your work and what's been done in the main repo, after you do the rebase you'll get merge conflicts that you need to fix.
